File: java/DominoAndTrominoTiling790.java
/*
You have two types of tiles: a 2 x 1 domino shape and a tromino shape. You may rotate these shapes.
Given an integer n, return the number of ways to tile an 2 x n board. Since the answer may be very large, return it modulo 109 + 7.
In a tiling, every square must be covered by a tile. Two tilings are different if and only if there are two 4-directionally adjacent cells on the board such that exactly one of the tilings has both squares occupied by a tile.
*/
public class DominoAndTrominoTiling790 {
    public static void main(String[] args) {
        int n = 3;
        System.out.println(numTilings(n));

        n = 4;
        System.out.println(numTilings(n));

        n = 5;
        System.out.println(numTilings(n));
    }

    public static int numTilings(int n) {
        if (n == 1) return 1;
        if (n == 2) return 2;
        if (n == 3) return 5;
        long long1 = 1, long2 = 2, long3 = 5, temp;
        for (int i = 4; i <= n; ++i) {
            temp = (2 * long3 + long1) % 1000000007;
            long1 = long2;
            long2 = long3;
            long3 = temp;
        }
        return (int) long3;

//        long[] g = new long[1001], u = new long[1001];
//        final int mod = 1000000007;
//        g[0] = 0;
//        g[1] = 1;
//        g[2] = 2;
//        u[0] = 0;
//        u[1] = 1;
//        u[2] = 2;
//        for (int i = 3; i <= n; i++) {
//            u[i] = (u[i - 1] + g[i - 1]) % mod;
//            g[i] = (g[i - 1] + g[i - 2] + 2 * u[i - 2]) % mod;
//        }
//        return (int) g[n] % mod;
    }
}

File: python/DominoAndTrominoTiling790.py
# You have two types of tiles: a 2 x 1 domino shape and a tromino shape. You may rotate these shapes. Given an
# integer n, return the number of ways to tile an 2 x n board. Since the answer may be very large, return it modulo
# 109 + 7. In a tiling, every square must be covered by a tile. Two tilings are different if and only if there are
# two 4-directionally adjacent cells on the board such that exactly one of the tilings has both squares occupied by a
# tile.
def numTilings(n: int) -> int:
    if n == 1:
        return 1
    if n == 2:
        return 2
    if n == 3:
        return 5
    one, two, three = 1, 2, 5
    for _ in range(4, n + 1):
        one, two, three = two, three, (2 * three + one) % 1000000007
    return three


if __name__ == "__main__":
    print(numTilings(3))  # Output: 5
    print(numTilings(4))  # Output: 11
    print(numTilings(5))  # Output: 24
    print(numTilings(6))  # Output: 53
    print(numTilings(7))  # Output: 117
    print(numTilings(8))  # Output: 258
    print(numTilings(9))  # Output: 569
    print(numTilings(10))  # Output: 1255
    print(numTilings(11))  # Output: 2767
    print(numTilings(12))  # Output: 6107
    print(numTilings(13))  # Output: 13475
    print(numTilings(14))  # Output: 29718
    print(numTilings(15))  # Output: 65533
    print(numTilings(16))  # Output: 144644
    print(numTilings(17))  # Output: 319122
    print(numTilings(18))  # Output: 704867
    print(numTilings(19))  # Output: 1552465
    print(numTilings(20))  # Output: 3423801
    print(numTilings(21))  # Output: 7549745
    print(numTilings(22))  # Output: 16641099
    print(numTilings(23))  # Output: 36629155
    print(numTilings(24))  # Output: 80702695
    print(numTilings(25))  # Output: 177997941
    print(numTilings(26))  # Output: 393069529
    print(numTilings(27))  # Output: 866988873
    print(numTilings(28))  # Output: 1917402657
    print(numTilings(29))  # Output: 4222475045
    print(numTilings(30))  # Output: 9319674297
    print(numTilings(31))  # Output: 205622
